src/core/connection.c

```diff
@@ -1892,6 +1892,223 @@ QuicConnRestart(
+        QUIC_TP_FLAG_INITIAL_MAX_STRMS_BIDI |
+        QUIC_TP_FLAG_INITIAL_MAX_STRMS_UNI);
+
+    const uint8_t* EncodedLocalTP = QuicCryptoTlsEncodeTransportParameters(
+        Connection,
+        &LocalTPCopy,
+        &EncodedTransportParametersLength);
+
+    if (EncodedLocalTP == NULL) {
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Error;
+    }
+    Ticket.TransportParamsLength = EncodedTransportParametersLength;
+
+    uint32_t TotalTicketLength = (uint32_t) (QuicVarIntSize(Ticket.TicketVersion) +
+        QuicVarIntSize(Ticket.AlpnLength) +
+        QuicVarIntSize(Ticket.TransportParamsLength) +
+        QuicVarIntSize(Ticket.AppTicketLength) +
+        sizeof(Ticket.QuicVersion) +
+        Ticket.AlpnLength +
+        Ticket.TransportParamsLength +
+        Ticket.AppTicketLength);
+
+    TicketBuffer = QUIC_ALLOC_NONPAGED(TotalTicketLength);
+    if (TicketBuffer == NULL) {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "Server resumption ticket",
+            TotalTicketLength);
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Error;
+    }
+
+    /*
+        Encoded ticket format is as follows:
+        Ticket Version (QUIC_VAR_INT) [1..4]
+        Quic Version [4]
+        Negotiated ALPN length (QUIC_VAR_INT) [1..4]
+        Negotiated ALPN [...]
+        Transport Parameters length (QUIC_VAR_INT) [1..4]
+        Transport Parameters [...]
+        App Ticket length (QUIC_VAR_INT) [1..4]
+        App Ticket (omitted if length is zero) [...]
+    */
+    uint8_t* TicketCursor = QuicVarIntEncode(Ticket.TicketVersion, TicketBuffer);
+    *(uint32_t*)TicketCursor = QuicByteSwapUint32(Ticket.QuicVersion);
+    TicketCursor += sizeof(Ticket.QuicVersion);
+    TicketCursor = QuicVarIntEncode(Ticket.AlpnLength, TicketCursor);
+    QuicCopyMemory(TicketCursor, Connection->Crypto.TlsState.NegotiatedAlpn + 1, (size_t) Ticket.AlpnLength);
+    TicketCursor += Ticket.AlpnLength;
+    TicketCursor = QuicVarIntEncode(Ticket.TransportParamsLength, TicketCursor);
+    QuicCopyMemory(TicketCursor, EncodedLocalTP, (size_t) Ticket.TransportParamsLength);
+    TicketCursor += EncodedTransportParametersLength;
+    TicketCursor = QuicVarIntEncode(Ticket.AppTicketLength, TicketCursor);
+    if (Ticket.AppTicketLength > 0) {
+        QuicCopyMemory(TicketCursor, AppResumptionData, (size_t) Ticket.AppTicketLength);
+    }
+
+    Status = QuicTlsSendTicket(
+        Connection->Crypto.TLS,
+        TotalTicketLength,
+        TicketBuffer);
+
+Error:
+    if (TicketBuffer != NULL) {
+        // This might need to be kept around for longer depending on TLS implementation...
+        QUIC_FREE(TicketBuffer);
+    }
+
+    if (EncodedLocalTP != NULL) {
+        QUIC_FREE(EncodedLocalTP);
+    }
+
+    if (AppResumptionData != NULL) {
+        QUIC_FREE(AppResumptionData);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
```

Comment:
I believe that in response to this call, we should actually be making a decision on whether or not we want to accept the resumption? What do you think? Does it still make sense for TLS to resume if there is something bad at our layer? I'm not sure...

Reply:
Should we fail the connection entirely, or just ignore resumption?

Reply:
For now, I am going to allow the app and QUIC to both reject the resumption data, and in response we'll reject all early data.